### PR TITLE
Implement basic status bar

### PR DESCRIPTION
**Pull request by astoeckel**
See #926. This adds a global status bar. It can be accessed using the following functions:

- Nengo.status_bar.set_caption()
- Nengo.status_bar.get_caption()
- Nengo.status_bar.hide()
- Nengo.status_bar.show()

Furthermore, this PR implements basic display of possible mouse/keyboard interactions with the object the mouse is currently over. Especially, it will tell you that you can
- Drag to pan if the mouse is over the netgraph
- Press CTRL to pan if over any object
- Drag to resize if over boundary of an object
- Drag to change slider value

Where the help string is adapted to the actual object type the mouse is over (graph, ensemble, node, network).

### CODE REVIEW

**Comment by tcstewar** [comment]
> Integration with #881 will be in that PR.

I would much prefer if the #881 PR could be merged without making it dependent on this one....

**Comment by astoeckel** [comment]
@tcstewar  Sure, I'll rebase this commit on #881 and implement the functionality here (because the status bar without any meaningful content probably doesn't make sense anyways).

File: nengo_gui/static/sim_control.css [hunk 1]
```diff
@@ -9,7 +9,8 @@
     flex: 0 0 auto;
     -webkit-flex: 0 0 auto;
     width: 100%;
-    z-index:99999999;
+    z-index:99999998; /* The SVG elements in the netgraph use z-index, so make
+                         sure they never cover the control panel */
```
Comment (jgosmann) [inline nengo_gui/static/sim_control.css]: It is easy to miss that this number ends in 8. Maybe point at that in the comment (and say why)?

File: nengo_gui/static/status_bar.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * Creates a global instance Nengo.status_bar inside the netgraph which can be
+ * accessed using the following functions:
+ * - Nengo.status_bar.set_caption()
+ * - Nengo.status_bar.get_caption()
+ * - Nengo.status_bar.hide()
+ * - Nengo.status_bar.show()
+ *
+ * @author Andreas Stöckel
+ * @date November 2017
```
Comment (jgosmann) [inline nengo_gui/static/status_bar.js]: We usually do not include authorship and date information in the source files. The VCS takes care of that information more accurately anyways.

File: nengo_gui/static/status_bar.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        "use strict";
+
+        // Create the component by manually assembling the DOM
+        this.status_bar_div = document.createElement('div');
+        this.status_bar_div.classList.add('status_bar');
+
+        // Timeout for updating the caption
+        this.caption_update_timeout = null;
+
+        // Append the component to the parent container
```
Comment (jgosmann) [inline nengo_gui/static/status_bar.js]: This comment doesn't really say anything, that the next line doesn't say already. I would remove it.

File: nengo_gui/static/status_bar.js (hunk 2 above)
Comment (astoeckel) [reply nengo_gui/static/status_bar.js]: Thanks, changed that.

File: nengo_gui/static/status_bar.js (hunk 3 above)
Comment (astoeckel) [reply nengo_gui/static/status_bar.js]: Removed last comment and reworded first comment

File: nengo_gui/static/sim_control.css (hunk 1 above)
Comment (astoeckel) [reply nengo_gui/static/sim_control.css]: I updated the comment accordingly. Would be nice to get rid of such magic numbers.

Comment (jgosmann) [reply nengo_gui/static/sim_control.css]: Yeah, but how?

Comment (jgosmann) [reply nengo_gui/static/sim_control.css]: 🤔

**Comment by astoeckel** [comment]
I've implemented a basic set of interaction displays in the status bar. As listed in the commit message, I'm displaying:

- Drag to pan if the mouse is over the netgraph
- Press CTRL to pan if over any object
- Drag to resize if over boundary of an object
- Drag to change slider value

This definitively improves the discoverability of the middle mouse button/CTRL pan action, and I also find it helpful that the object type, e.g. 'node', 'ensemble', 'graph', etc. is shown in the status bar. In the future, once it is actually possible to select objects, it might make sense to show information such as '3 nodes, 2 ensembles (100 neurons) selected.' (compare Inkscape).

As for the position of the status bar I'm not entirely sure whether it should stay where it is. In case it is too distractive where it is right now, it might also be moved to the very bottom of the window.

Tested with Firefox and Chromium on Linux.

Feedback welcome!

File: nengo_gui/static/sim_control.css (hunk 1 above)
Comment (astoeckel) [reply nengo_gui/static/sim_control.css]: Just checked the DOM again. I don't see any need for the z-index at all (the comment is based on what Terry told me). The netgraph is contained within a DOM element with `overflow: hidden` and there are no elements with `position: fixed`, so these are two completely separate object stacks.

**Comment by jgosmann** [comment]
> This definitively improves the discoverability of the middle mouse button/CTRL pan action,

👍 

> I also find it helpful that the object type, e.g. 'node', 'ensemble', 'graph', etc., is shown in the status bar

I don't find that information very salient because it is hidden in the sentence. Putting it at the front of the sentence (“Node: [you can do these things with it]”) or adding some styling (bold) to it might help.

> As for the position of the status bar I'm not entirely sure whether it should stay where it is. In case it is too distractive where it is right now, it might also be moved to the very bottom of the window.

I'd move it to the bottom of the window (though I have to check what that look like). It is also cut-off by the console at the right it the moment which looks super weird to me.

**Comment by astoeckel** [comment]
Thanks for the feedback, I'll try to incorporate it into a next revision as soon as I find the time (which could take two weeks).

**Comment by jgosmann** [comment]
@astoeckel Do you still intend to work on this? I was thinking about integrating the build progress bar with Nengo GUI and a status bar might be the logical place to put it.

**Comment by jgosmann** [comment]
I added my proposed changes. I think this should be ready for review now?

**Comment by tcstewar** [comment]
This has now been merged as part of #958   Thank you @astoeckel for doing this!  I really like how it turned out!

**Pull request closed without merging**